File: src/jvm/clojure/lang/APersistentMap.java
/**
 *   Copyright (c) Rich Hickey. All rights reserved.
 *   The use and distribution terms for this software are covered by the
 *   Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
 *   which can be found in the file epl-v10.html at the root of this distribution.
 *   By using this software in any fashion, you are agreeing to be bound by
 * 	 the terms of this license.
 *   You must not remove this notice, or any other, from this software.
 **/

package clojure.lang;

import java.io.Serializable;
import java.util.*;

public abstract class APersistentMap extends AFn implements IPersistentMap, Map, Iterable, Serializable, MapEquivalence, IHashEq {
int _hash;
int _hasheq;

public String toString(){
	return RT.printString(this);
}

public IPersistentCollection cons(Object o){
	if(o instanceof Map.Entry)
		{
		Map.Entry e = (Map.Entry) o;

		return assoc(e.getKey(), e.getValue());
		}
	else if(o instanceof IPersistentVector)
		{
		IPersistentVector v = (IPersistentVector) o;
		if(v.count() != 2)
			throw new IllegalArgumentException("Vector arg to map conj must be a pair");
		return assoc(v.nth(0), v.nth(1));
		}

	IPersistentMap ret = this;
	for(ISeq es = RT.seq(o); es != null; es = es.next())
		{
		Map.Entry e = (Map.Entry) es.first();
		ret = ret.assoc(e.getKey(), e.getValue());
		}
	return ret;
}

public boolean equals(Object obj){
	return mapEquals(this, obj);
}

static public boolean mapEquals(IPersistentMap m1, Object obj){
	if(m1 == obj) return true;
	if(!(obj instanceof Map))
		return false;
	Map m = (Map) obj;

	if(m.size() != m1.count())
		return false;

	for(ISeq s = m1.seq(); s != null; s = s.next())
		{
		Map.Entry e = (Map.Entry) s.first();
		boolean found = m.containsKey(e.getKey());

		if(!found || !Util.equals(e.getValue(), m.get(e.getKey())))
			return false;
		}

	return true;
}

public boolean equiv(Object obj){
	if(!(obj instanceof Map))
		return false;
	if(obj instanceof IPersistentMap && !(obj instanceof MapEquivalence))
		return false;
	
	Map m = (Map) obj;

	if(m.size() != size())
		return false;

	for(ISeq s = seq(); s != null; s = s.next())
		{
		Map.Entry e = (Map.Entry) s.first();
		boolean found = m.containsKey(e.getKey());

		if(!found || !Util.equiv(e.getValue(), m.get(e.getKey())))
			return false;
		}

	return true;
}
public int hashCode(){
    int cached = this._hash;
	if(cached == 0)
		{
		this._hash = cached = mapHash(this);
		}
	return cached;
}

static public int mapHash(IPersistentMap m){
	int hash = 0;
	for(ISeq s = m.seq(); s != null; s = s.next())
		{
		Map.Entry e = (Map.Entry) s.first();
		hash += (e.getKey() == null ? 0 : e.getKey().hashCode()) ^
				(e.getValue() == null ? 0 : e.getValue().hashCode());
		}
	return hash;
}

public int hasheq(){
    int cached = this._hasheq;
	if(cached == 0)
		{
		//this._hasheq = mapHasheq(this);
		this._hasheq = cached = Murmur3.hashUnordered(this);
		}
	return cached;
}

static public int mapHasheq(IPersistentMap m) {
	return Murmur3.hashUnordered(m);
//	int hash = 0;
//	for(ISeq s = m.seq(); s != null; s = s.next())
//		{
//		Map.Entry e = (Map.Entry) s.first();
//		hash += Util.hasheq(e.getKey()) ^
//				Util.hasheq(e.getValue());
//		}
//	return hash;
}

static public class KeySeq extends ASeq{
	final ISeq seq;
	final Iterable iterable;

	static public KeySeq create(ISeq seq){
		if(seq == null)
			return null;
		return new KeySeq(seq, null);
	}

	static public KeySeq createFromMap(IPersistentMap map){
		if(map == null)
			return null;
		ISeq seq = map.seq();
		if(seq == null)
			return null;
		return new KeySeq(seq, map);
	}

	private KeySeq(ISeq seq, Iterable iterable){
		this.seq = seq;
		this.iterable = iterable;
	}

	private KeySeq(IPersistentMap meta, ISeq seq, Iterable iterable){
		super(meta);
		this.seq = seq;
		this.iterable = iterable;
	}

	public Object first(){
		return ((Map.Entry) seq.first()).getKey();
	}

	public ISeq next(){
		return create(seq.next());
	}

	public KeySeq withMeta(IPersistentMap meta){
		if(meta() == meta)
			return this;
		return new KeySeq(meta, seq, iterable);
	}

	public Iterator iterator(){
		if(iterable == null)
			return super.iterator();

		if(iterable instanceof IMapIterable)
			return ((IMapIterable)iterable).keyIterator();

		final Iterator mapIter = iterable.iterator();
		return new Iterator() {
			public boolean hasNext() {
				return mapIter.hasNext();
			}

			public Object next() {
				return ((Map.Entry)mapIter.next()).getKey();
			}

			public void remove() {
				throw new UnsupportedOperationException();
			}
		};
	}
}

static public class ValSeq extends ASeq{
	final ISeq seq;
	final Iterable iterable;

	static public ValSeq create(ISeq seq){
		if(seq == null)
			return null;
		return new ValSeq(seq, null);
	}

	static public ValSeq createFromMap(IPersistentMap map) {
		if(map == null)
			return null;
		ISeq seq = map.seq();
		if(seq == null)
			return null;
		return new ValSeq(seq, map);
	}

	private ValSeq(ISeq seq, Iterable iterable){
		this.seq = seq;
		this.iterable = iterable;
	}

	private ValSeq(IPersistentMap meta, ISeq seq, Iterable iterable){
		super(meta);
		this.seq = seq;
		this.iterable = iterable;
	}

	public Object first(){
		return ((Map.Entry) seq.first()).getValue();
	}

	public ISeq next(){
		return create(seq.next());
	}

	public ValSeq withMeta(IPersistentMap meta){
		if(meta() == meta)
			return this;
		return new ValSeq(meta, seq, iterable);
	}

	public Iterator iterator(){
		if(iterable == null)
			return super.iterator();

		if(iterable instanceof IMapIterable)
			return ((IMapIterable)iterable).valIterator();

		final Iterator mapIter = iterable.iterator();
		return new Iterator() {
			public boolean hasNext() {
				return mapIter.hasNext();
			}

			public Object next() {
				return ((Map.Entry)mapIter.next()).getValue();
			}

			public void remove() {
				throw new UnsupportedOperationException();
			}
		};
	}
}

static final IFn MAKE_ENTRY = new AFn() {
    public Object invoke(Object key, Object val) {
        return MapEntry.create(key, val);
    }
};

static final IFn MAKE_KEY = new AFn() {
    public Object invoke(Object key, Object val) {
        return key;
    }
};

static final IFn MAKE_VAL = new AFn() {
    public Object invoke(Object key, Object val) {
        return val;
    }
};

public Object invoke(Object arg1) {
	return valAt(arg1);
}

public Object invoke(Object arg1, Object notFound) {
	return valAt(arg1, notFound);
}

// java.util.Map implementation

public void clear(){
	throw new UnsupportedOperationException();
}

public boolean containsValue(Object value){
	return values().contains(value);
}

public Set entrySet(){
	return new AbstractSet(){

		public Iterator iterator(){
			return APersistentMap.this.iterator();
		}

		public int size(){
			return count();
		}

		public int hashCode(){
			return APersistentMap.this.hashCode();
		}

		public boolean contains(Object o){
			if(o instanceof Entry)
				{
				Entry e = (Entry) o;
				Entry found = entryAt(e.getKey());
				if(found != null && Util.equals(found.getValue(), e.getValue()))
					return true;
				}
			return false;
		}
	};
}

public Object get(Object key){
	return valAt(key);
}

public boolean isEmpty(){
	return count() == 0;
}

public Set keySet(){
	return new AbstractSet(){

		public Iterator iterator(){
			final Iterator mi = APersistentMap.this.iterator();

			return new Iterator(){


				public boolean hasNext(){
					return mi.hasNext();
				}

				public Object next(){
					Entry e = (Entry) mi.next();
					return e.getKey();
				}

				public void remove(){
					throw new UnsupportedOperationException();
				}
			};
		}

		public int size(){
			return count();
		}

		public boolean contains(Object o){
			return APersistentMap.this.containsKey(o);
		}
	};
}

public Object put(Object key, Object value){
	throw new UnsupportedOperationException();
}

public void putAll(Map t){
	throw new UnsupportedOperationException();
}

public Object remove(Object key){
	throw new UnsupportedOperationException();
}

public int size(){
	return count();
}

public Collection values(){
	return new AbstractCollection(){

		public Iterator iterator(){
			final Iterator mi = APersistentMap.this.iterator();

			return new Iterator(){


				public boolean hasNext(){
					return mi.hasNext();
				}

				public Object next(){
					Entry e = (Entry) mi.next();
					return e.getValue();
				}

				public void remove(){
					throw new UnsupportedOperationException();
				}
			};
		}

		public int size(){
			return count();
		}
	};
}

/*
// java.util.Collection implementation

public Object[] toArray(){
	return RT.seqToArray(seq());
}

public boolean add(Object o){
	throw new UnsupportedOperationException();
}

public boolean remove(Object o){
	throw new UnsupportedOperationException();
}

public boolean addAll(Collection c){
	throw new UnsupportedOperationException();
}

public void clear(){
	throw new UnsupportedOperationException();
}

public boolean retainAll(Collection c){
	throw new UnsupportedOperationException();
}

public boolean removeAll(Collection c){
	throw new UnsupportedOperationException();
}

public boolean containsAll(Collection c){
	for(Object o : c)
		{
		if(!contains(o))
			return false;
		}
	return true;
}

public Object[] toArray(Object[] a){
	if(a.length >= count())
		{
		ISeq s = seq();
		for(int i = 0; s != null; ++i, s = s.rest())
			{
			a[i] = s.first();
			}
		if(a.length > count())
			a[count()] = null;
		return a;
		}
	else
		return toArray();
}

public int size(){
	return count();
}

public boolean isEmpty(){
	return count() == 0;
}

public boolean contains(Object o){
	if(o instanceof Map.Entry)
		{
		Map.Entry e = (Map.Entry) o;
		Map.Entry v = entryAt(e.getKey());
		return (v != null && Util.equal(v.getValue(), e.getValue()));
		}
	return false;
}
*/
}

File: ClojureCLR/Clojure/Clojure/Lib/APersistentMap.cs
/**
 *   Copyright (c) David Miller. All rights reserved.
 *   The use and distribution terms for this software are covered by the
 *   Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
 *   which can be found in the file epl-v10.html at the root of this distribution.
 *   By using this software in any fashion, you are agreeing to be bound by
 * 	 the terms of this license.
 *   You must not remove this notice, or any other, from this software.
 **/

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Reflection;

using System.Collections;

namespace clojure.lang
{
    /// <summary>
    /// Provides a basic impelmentation of <see cref="IPersistentMap">IPersistentMap</see> functionality.
    /// </summary>
    public abstract class APersistentMap: AFn, IPersistentMap, IDictionary, IEnumerable<IMapEntry>
    {
        #region  Data
        
        /// <summary>
        /// Caches the hash code, when computed.
        /// </summary>
        /// <remarks>The value <value>-1</value> indicates that the hash code has not been computed yet.</remarks>
        int _hash = -1;

        #endregion

        #region C-tors and factory methods

        /// <summary>
        /// Initializes a <see cref="APersistnetMap">APersistantMap</see> to have the given metadata.
        /// </summary>
        /// <param name="meta">The metatdata to attach.</param>
        protected APersistentMap(IPersistentMap meta)
            : base(meta)
        {
        }

        /// <summary>
        /// Initializes a <see cref="APersistnetMap">APersistantMap</see> to have null metadata.
        /// </summary>
        protected APersistentMap()
        {
        }

        #endregion

        #region object overrides

        /// <summary>
        /// Returns a string that represents the current object.
        /// </summary>
        /// <returns>A string that represents the current object.</returns>
        public override string ToString()
        {
            return RT.printString(this);
        }

        /// <summary>
        /// Determines whether the specified Object is equal to the current Object.
        /// </summary>
        /// <param name="obj">The Object to compare with the current Object. </param>
        /// <returns>true if the specified Object is equal to the current Object; 
        /// otherwise, false.</returns>
        public override bool Equals(object obj)
        {
            //if(!(obj instanceof Map))
            //    return false;
            //Map m = (Map) obj;

            IDictionary d = obj as IDictionary;
            if (d == null)
                return false;

            // Java had the following.
            // This works on other APersistentMap implementations, but not on
            //  arbitrary dictionaries.
            if (d.Count != this.Count || d.GetHashCode() != this.GetHashCode())
                return false;

            //if (d.Count != this.Count)
            //    return false;

            for (ISeq s = seq(); s != null; s = s.next())
            {
                IMapEntry me = (IMapEntry)s.first();
                bool found = d.Contains(me.key());
                if (!found || !Util.equals(me.val(), d[me.key()]))
                    return false;
            }
            return true;
        }

 
        /// <summary>
        /// Gets a hash code for the current object.
        /// </summary>
        /// <returns>A hash code for the current object.</returns>
        /// <remarks>Valud-based = relies on all entries.  Once computed, it is cached.</remarks>
        public override int GetHashCode()
        {
            if (_hash == -1)
            {
                int hash = 0;
                for (ISeq s = seq(); s != null; s = s.next())
                {
                    IMapEntry me = (IMapEntry)s.first();
                    hash += (me.key() == null ? 0 : me.key().GetHashCode())
                        ^ (me.val() == null ? 0 : me.val().GetHashCode());
                }
                _hash = hash;
            }
            return _hash;
        }


        #endregion

        #region Associative methods

        abstract public bool containsKey(object key);
        abstract public IMapEntry entryAt(object key);
        Associative Associative.assoc(object key, object val)
        {
            return assoc(key, val);
        }
        abstract public object valAt(object key);
        abstract public object valAt(object key, object notFound);

        #endregion

        #region Seqable members

        abstract public ISeq seq();

        #endregion

        #region IPersistentCollection Members

        /// <summary>
        /// Returns a new collection that has the given element cons'd on front of the eixsting collection.
        /// </summary>
        /// <param name="o">An item to put at the front of the collection.</param>
        /// <returns>A new immutable collection with the item added.</returns>
        IPersistentCollection IPersistentCollection.cons(object o)
        {
            return cons(o);
        }

        abstract public int count();
        abstract public IPersistentCollection empty();

        /// <summary>
        /// Determine if an object is equivalent to this (handles all collections).
        /// </summary>
        /// <param name="o">The object to compare.</param>
        /// <returns><c>true</c> if the object is equivalent; <c>false</c> otherwise.</returns>
        /// <remarks>
        /// In Java Rev 1215, Added equiv.  Same as the definition in Equals, as in they took out the hashcode comparison.
        /// Different, as in Util.Equal above became Util.equals. and below it is Util.equiv.
        /// </remarks> 
        public bool equiv(object obj)
        {
            //if(!(obj instanceof Map))
            //    return false;
            //Map m = (Map) obj;

            IDictionary d = obj as IDictionary;
            if (d == null)
                return false;

            // Java had the following.
            // This works on other APersistentMap implementations, but not on
            //  arbitrary dictionaries.
            //if (d.Count != this.Count || d.GetHashCode() != this.GetHashCode())
            //    return false;

            if (d.Count != this.Count)
                return false;

            for (ISeq s = seq(); s != null; s = s.next())
            {
                IMapEntry me = (IMapEntry)s.first();
                bool found = d.Contains(me.key());
                if (!found || !Util.equiv(me.val(), d[me.key()]))
                    return false;
            }
            return true;
        }



        #endregion

        #region IObj members

        abstract public override IObj withMeta(IPersistentMap meta);

        #endregion

        #region IPersistentMap members

        abstract public IPersistentMap assoc(object key, object val);
        abstract public IPersistentMap assocEx(object key, object val);
        abstract public IPersistentMap without(object key);

        /// <summary>
        /// Add a new key/value pair.
        /// </summary>
        /// <param name="o">The key/value pair to add.</param>
        /// <returns>A new map with key+value pair added.</returns>
        public IPersistentMap cons(object o)
        {
            IMapEntry e = o as IMapEntry;
            if (e != null)
                return assoc(e.key(), e.val());

            if (o is DictionaryEntry)
            {
                DictionaryEntry de = (DictionaryEntry)o;
                return assoc(de.Key, de.Value);
            }
            Type t = o.GetType();
            if (t.IsGenericType && t.Name == "KeyValuePair`2")
            {
                object key = t.InvokeMember("Key", BindingFlags.GetProperty, null, o, null);
                object val = t.InvokeMember("Value", BindingFlags.GetProperty, null, o, null);
                return assoc(key, val);
            }

            IPersistentVector v = o as IPersistentVector;
            if (v != null)
            {
                if (v.count() != 2)
                    throw new ArgumentException("Vector arg to map cons must be a pair");
                return assoc(v.nth(0), v.nth(1));
            }

            IPersistentMap ret = this;
            for (ISeq s = RT.seq(o); s != null; s = s.next())
            {
                IMapEntry me = (IMapEntry)s.first();
                ret = ret.assoc(me.key(), me.val());
            }
            return ret;
        }

        #endregion

        #region IFn members

        public override object invoke(object arg1)
        {
            return valAt(arg1);
        }

        public override object invoke(object arg1, object arg2)
        {
            return valAt(arg1, arg2);
        }

        #endregion

        #region IDictionary Members

        public void Add(object key, object value)
        {
            throw new NotImplementedException();
        }

        public void Clear()
        {
            throw new NotImplementedException();
        }

        public bool Contains(object key)
        {
            return this.containsKey(key);
        }

        public virtual IDictionaryEnumerator GetEnumerator()
        {
            return new MapEnumerator(this);
        }

        public bool IsFixedSize
        {
            get { return true; }
        }

        public bool IsReadOnly
        {
            get { return true; }
        }

        public ICollection Keys
        {
            get { return KeySeq.create(seq()); }
        }

        public void Remove(object key)
        {
            throw new NotImplementedException();
        }

        public ICollection Values
        {
            get { return ValSeq.create(seq()); }
        }

        public object this[object key]
        {
            get
            {
                return valAt(key);
            }
            set
            {
                throw new NotImplementedException();
            }
        }

        #endregion

        #region ICollection Members

        public void CopyTo(Array array, int index)
        {
            ((ICollection)seq()).CopyTo(array, index);
        }

        public int Count
        {
            get { return count(); }
        }

        public bool IsSynchronized
        {
            get { return true; }
        }

        public object SyncRoot
        {
            get { throw new NotImplementedException(); }
        }

        #endregion

        #region IEnumerable Members

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        #endregion

        /// <summary>
        /// Implements a sequence across the keys of map.
        /// </summary>
        public sealed class KeySeq : ASeq
        {
            #region Data

            ISeq _seq;

            #endregion

            #region C-tors & factory methods

            static public KeySeq create(ISeq seq)
            {
                if (seq == null)
                    return null;
                return new KeySeq(seq);
            }

            private KeySeq(ISeq seq)
            {
                _seq = seq;
            }

            private KeySeq(IPersistentMap meta, ISeq seq)
                : base(meta)
            {
                _seq = seq;
            }

            #endregion

            #region ISeq members

            public override object first()
            {
                return ((IMapEntry)_seq.first()).key();
            }

            public override ISeq next()
            {
                return create(_seq.next());
            }

            #endregion

            #region IObj methods

            public override IObj withMeta(IPersistentMap meta)
            {
                return new KeySeq(meta, _seq);
            }

            #endregion
        }

        /// <summary>
        /// Implements a sequence across the values of a map.
        /// </summary>
        public sealed class ValSeq : ASeq
        {
            #region Data

            ISeq _seq;

            #endregion

            #region C-tors & factory methods

            static public ValSeq create(ISeq seq)
            {
                if (seq == null)
                    return null;
                return new ValSeq(seq);
            }

            private ValSeq(ISeq seq)
            {
                _seq = seq;
            }

            private ValSeq(IPersistentMap meta, ISeq seq)
                : base(meta)
            {
                _seq = seq;
            }

            #endregion

            #region ISeq members

            public override object first()
            {
                return ((IMapEntry)_seq.first()).val();
            }

            public override ISeq next()
            {
                return create(_seq.next());
            }

            #endregion

            #region IObj methods

            public override IObj withMeta(IPersistentMap meta)
            {
                return new ValSeq(meta, _seq);
            }

            #endregion
        }


        #region IEnumerable<IMapEntry> Members

        IEnumerator<IMapEntry> IEnumerable<IMapEntry>.GetEnumerator()
        {
            for (ISeq s = seq(); s != null; s = s.next())
                yield return (IMapEntry)s.first();
        }

        #endregion
    }


}
